Adding OrchRefs to startup

heschlie opened the pull request:
## Description

Adding in OrchRefs to startup.go, and revving libcalico-go to the required commit.

I've got k8s hardcoded as the "Orchestrator" right now, which feels wrong, seeking advice on how we should be setting it.

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

<!-- Writing a release note:
- By default, no release note action is required.
- If you're unsure whether or not your PR needs a note, ask your reviewer for guidance.
- If this PR requires a release note, update the block below to include a concise note describing
  the change and any important impacts this PR may have.
-->

```release-note
None required
```

caseydavenport commented:
@heschlie not sure what's going on with the build, looks like you've got some dependency fun going on.

heschlie commented:
I _think_ one or more of the glide pins broke something last night, reran `glide up -v` and it passed the `node-test-containerized` locally.

heschlie commented:
Seeing some calicoctl related failures with `config` not being a valid option, need to set it to an older version of calicoctl.

heschlie commented:
It looks like the `make semaphore` target is hard coded to `RELEASE_STREAM=master`. I've set it to `RELEASE_STREAM=$(RELEASE_STREAM)` and am running the tests again as I was still seeing some failure after setting calicoctl versions to v1.6.1.

ozdanborne commented on calico_node/glide.yaml:
```diff
@@ -3,7 +3,7 @@ import:
   version: ^0.10.0
 - package: github.com/projectcalico/libcalico-go
-  version: v1.7.1
+  version: 6410b9a3ddf7e3b2e33e1e9acd7984e00101ec8f
```
should this just be `v1.x-series`?

heschlie replied:
It certainly could be, but that is technically a moving target isn't it (though a very slow one)

caseydavenport replied:
It's fine to have moving targets in glide.yaml

It's the glide.lock file that will store a specific commit.

caseydavenport replied:
That said, either a commit or a branch is fine.

heschlie replied:
I feel like the branch is easier to grok when looking at the file. I assume it'll get pinned to a release tag during a `v2.x-series` release anyway.

heschlie commented:
Hitting some errors with AOF tests:

```
CommandExecError: Command docker exec -it cali-st-gw sh -c 'export ETCD_AUTHORITY=172.24.16.173:2379; export ETCD_SCHEME=http; export ETCD_CA_CERT_FILE=; export ETCD_CERT_FILE=; export ETCD_KEY_FILE=; /code/dist/calicoctl apply -f new_data' failed with RC 1 and output:
Failed to execute command: error parsing document: field in document is not recognized or is in the wrong location: applyOnForward
```

Is there a particular version of calicoctl we should be targeting? It looks like it should be using `v1.6.1` if I'm understanding things correctly.

heschlie commented:
@caseydavenport I think AOF might be missing from calicoctl v1.6.1, I copied the config from the tests that do AOF and tried to apply it manually and am seeing the same failure:

yaml:
```yaml
apiVersion: v1
kind: policy
metadata:
  name: host-out
spec:
  order: 100
  selector: nodeEth == "host"
  types:
    - egress
    - ingress
  egress:
    - action: allow
  ingress:
    - action: allow
  applyOnForward: true
```

calicoctl version:
```
Client Version:    v1.6.1
Build date:        2017-09-28T01:12:35+0000
Git commit:        1724e011
Cluster Version:   unknown
Cluster Type:      unknown
```

Is this expected? I can change the tests if so, but looking through the commit history I think it is supposed to be supported.

heschlie commented:
I've fixed the issues locally now after cutting a new branch for calicoctl for the `v1.x-series` and updating the glide pins there (PR projectcalico/calicoctl#1745) so we're in the home stretch.

With this branch for Calico node `v2.x-series` do we want to change the versions in the master context of the `versions.yaml` to match all of the appropriate branches that reflect the "master" of Calico v2?

caseydavenport commented:
@heschlie hm, interesting point.

Yes, I think we probably do given this is going into the v2.x series branch. @ozdanborne for confirmation.

heschlie commented:
Just realized that `0e5af58` is going to fail because we need to wait for a new version of the calicoctl `v1.x-series` images to be pushed after projectcalico/calicoctl#1745 is merged.

heschlie commented:
Seeing errors in the htmlproofer, were those links removed and should be fixed here as well?

caseydavenport commented:
@heschlie yeah, those links were removed in master. 

I think we should add those to the list of ignored files.

caseydavenport commented:
Given we're never building the docs from this branch - it's just for maintaining a v2.x calico/node - we can probably just turn off htmlproofer entirely. 

i.e. there should never be docs PRs to this branch.

heschlie commented:
I'm not sure how easy it would be to disable it against this branch in Semaphore (not sure that is even what you are suggesting) but I can just replace the make target with a message that says don't make docs changes here, do it in master.

caseydavenport commented:
@heschlie SGTM

caseydavenport commented:
@heschlie are you planning on adding an ST to test this?

ozdanborne commented on calico_node/startup/startup.go:
```diff
@@ -102,6 +102,13 @@ func main() {
 	}
 
+	if nodeRef := os.Getenv("CALICO_K8S_NODE_REF"); nodeRef != "" {
+		node.Spec.OrchRefs = append(node.Spec.OrchRefs, api.OrchRef{
```
Doesn't this mean that each time the node is rebooted, it will add an additional orchRef? 

I think it needs to check if a "k8s" orchRef has been set yet, and if not set it.

Additionally, I think we need to detect which orchestrator we're running in here. Looks like this code path runs for all orchestrators. But we only want to set it if we're running in k8s.

heschlie replied:
Yea I agree on both points, do we have a good way to detect the orchestrator?

heschlie commented:
@ozdanborne @caseydavenport I've fixed up the logic a bit and moved it into its own function so I could make some tests to hammer against it. There is a table driven test in there I added for this, that only has one entry right now, but is there to easily add to it when we add more orchestrators into this.

heschlie commented on calico_node/startup/startup.go:
```diff
@@ -128,6 +130,31 @@ func main() {
+		NodeName:     nodeRef,
+		Orchestrator: orchestrator,
+	}
+	// Return if we find that this OrchRef is not already set on the Node.
```
s/is not/not

caseydavenport commented on calico_node/startup/startup.go:
```diff
@@ -128,6 +130,31 @@ func main() {
+	orchestrator := ""
+	nodeRef := ""
+	// Sort out what type of cluster we're running on.
+	if nodeRef = os.Getenv("CALICO_K8S_NODE_REF"); nodeRef != "" {
```
This could be simplified, since we only support this feature on k8s right now.

```go
orch := "k8s"
nodeRef := ""
if nodeRef = os.Getenv("CALICO_K8S_NODE_REF"); nodeRef == "" {
  // No reference specified, return.
  return
}
```

heschlie replied:
yea, in my mind I was thinking "We will probably want to check for other orchestrator env vars." but I don't know what we plan to set for their env vars or how imminent their implementation will be.

caseydavenport replied:
They're not planned any time soon, so I'd opt for the simpler approach here.

We'll cross that bridge when we get to it - generally best to keep things as simple as possible until required.

caseydavenport commented on calico_node/startup/startup.go:
```diff
@@ -128,6 +130,31 @@ func main() {
+		Orchestrator: orchestrator,
+	}
+	// Return if we find that this OrchRef is not already set on the Node.
+	for _, ref := range node.Spec.OrchRefs {
```
Probably fine, but given we know we only support this on k8s we could do a simple `if len(node.Spec.OrchRefs) != 0`.

Don't care too much either way.

caseydavenport replied:
A simpler approach might just be to write this directly every time.

This is slightly different, in that we'll actually update the orch ref if it changes. Unclear if that is more or less desirable. I think it is probably better.

```go
node.Spec.OrchRefs = []api.OrchRef{api.OrchRef{. . .}}
```

heschlie replied:
I think I agree, it would make it a bit more straightforward to expand on later, but I don't know if there are any repercussions if we inadvertently remove some OrchRefs, but I don't _think_ that is a problem.

caseydavenport commented on calico_node/startup/startup.go:
```diff
@@ -128,6 +130,31 @@ func main() {
+func configureNodeRef(node *api.Node) {
+	orchestrator := ""
+	nodeRef := ""
+	// Sort out what type of cluster we're running on.
```
nit - should have a newline before the comments in this fucntion

ozdanborne commented on Makefile:
```diff
@@ -32,12 +32,8 @@ _site build:
-	
-	docker run -v $$PWD:/calico --entrypoint /bin/sh -ti garethr/kubeval:0.1.1 -c 'find /calico/_site/master -name "*.yaml" |grep -v config.yaml | xargs /kubeval'
+htmlproofer:
+	@echo "Do not make docs changes against this branch, please use master."
```
What's the reasoning behind this?

caseydavenport replied:
@ozdanborne this is because the htmlproofer checks bunch of github docs that no longer exist in the master branch because master is tracking v3.0.

The docs will always be rendered from the master branch anyway - this branch is just for maintaining a calico/node build, so we decided just to disable the htmlproofer check.

heschlie replied:
Caseys comment from earlier:
>Given we're never building the docs from this branch - it's just for maintaining a v2.x calico/node >- we can probably just turn off htmlproofer entirely.
>
>i.e. there should never be docs PRs to this branch.

So I just replaced it with a message so we don't have to do anything clever in semaphore to git it working against this branch.
